File: src/4.单例模式/demo1.java

public class SingleObject {
    // 注意，私有化构造函数，外部不能new，只有内部能 new ！！！
    private SingleObject(){
    }
    // 唯一被 new 出来的对象
    private SingleObject instance = null
    // 唯一被 new 出来的对象
    public SingleObject getInstance(){
        if(instance == null){
            // 只 new 一次
            instance = new SingleObject()
        }
        return instance
    }

    // 对象方法
    public void login(username,password){
        System.out.printIn('login...')
    }
}

public class SingletonPatternDemo{
    public static void main(String[] args){
        // 不合法的构造函数
        // 编译时错误：构造 SingleObject() 是不可见的!!!
        // SingleObject object = new SingleObject()

        // 获取唯一可用的对象
        SingleObject object = SingleObject.getInstance()
        object.login()
    }
}

File: src/4.单例模式/demo1.js
class SingleObject {
    login() {
        console.log('login...')
    }
}
// 静态方法 无论 SingleObject 被 new多少个，这个getInstance方法也只有一个
SingleObject.getInstance = (function () {
    let instance
    return function () {
        if (!instance) {
            instance = new SingleObject();
        }
        return instance
    }
})()

// 测试:主义这里只能使用静态函数 getInstance，不能 new SingleObject()  ！！！
let obj1 = SingleObject.getInstance()
obj1.login()
let obj2 = SingleObject.getInstance()
obj2.login()
console.log(obj1 === obj2)

console.log(`-------------`);

let obj3 = new SingleObject() // 无法完全控制
obj3.login()
console.log(obj1 === obj3);